### PR TITLE
Fix HTTP/2 peer_data

### PR DESCRIPTION
**Pull request by mtrudel**
Also take the change to refactor some of the plumbing.

Contrast with #171

### CODE REVIEW

**Comment by moogle19** [comment]
Could be move `build_transport_info/1` into the `Bandit.SocketHelpers` module? 
Currently the code is duplicated in `HTTP1.Adapter` and `HTTP2.Connection`.

**Comment by mtrudel** [comment]
Ohhhhh that's a good idea!

**Comment by moogle19** [comment]
Also `build_transport_info/1`and `peer_data/1` both call `ThousandIsland.Socket.peername/1`.
Is there a good way to reuse the transport_info when constructing the `peer_data`? 
Maybe even split up `transport_info` since at most one element of the tuple is used in a function.

And it looks like the `Telemetry` part on the transport_info is never used after initial creation.

**Comment by mtrudel** [comment]
I've got a WiP on the go here, but I'm not going to be able to finish it today. I've already got the general idea sketched out; hopefully I'll get a spike up by this evening.

**Comment by mtrudel** [comment]
It's up at https://github.com/mtrudel/bandit/commit/27c2d2cee157d13968c22cd0c1c5ccef3133709b. The general idea is to abstract all aspects of transport info into a module that inits itself directly off of a socket, and then answer all questions about transport info (whether for Plug adapters or in the internals of the pipeline) from there. 

The HTTP/2 side still needs to be done, and there are a few TODOs pointed out in code, but the idea hangs together really well I think.

**Comment by mtrudel** [comment]
HTTP/2 ready to go. Barring two TODOs in Pipeline, I think this is ready to stand for review

**Review by moogle19 — APPROVED** [review]
This looks awesome! 
It makes the code much easier to follow and removes a good bunch of duplication.

File: lib/bandit/transport_info.ex [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    end
+  end
+
+  @spec peercert(ThousandIsland.Socket.t()) ::
+          {:ok, :public_key.der_encoded() | nil} | {:error, term()}
+  defp peercert(socket) do
+    case ThousandIsland.Socket.peercert(socket) do
+      {:ok, cert} -> {:ok, cert}
+      {:error, :no_peercert} -> {:ok, nil}
+      {:error, :not_secure} -> {:ok, nil}
```
[the remark this answers is not in the document]
Comment (mtrudel) [reply lib/bandit/transport_info.ex:45]: They're straight up different sources (one from TCP, one from SSL) which is why I didn't DRY them up.

**Pull request merged**